std/assembly/array.ts

```diff
@@ -476,6 +476,64 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 1 : child.length;
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    let byteLength = size << usize(alignof<valueof<T>>());
+    let dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value
+    let result: valueof<T>[] = changetype<valueof<T>[]>(__alloc(offsetof<valueof<T>[]>(), idof<valueof<T>[]>()));
+    store<i32>(changetype<usize>(result), size, offsetof<valueof<T>[]>("length_"));
+    result.dataLength = byteLength;
+    result.dataStart = dataStart;
+    // this statement retains the ArrayBuffer
+    result.data = changetype<ArrayBuffer>(dataStart);
+
+    // set the elements
+    let resultIndex: i32 = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      resultIndex++;
+
+      if (child == null) {
+        // if child == null and valueof<T> is a numeric type, it will push 0 here
+        store<valueof<T>>(dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())), null);
```

Comment:
Hmm, wondering what happens if `valueof<T>` is non-nullable here. Seems the return type of the function would be `string[]` for example with interior `null`s, which isn't type safe anymore. Might have to throw `E_HOLEYARRAY` in this case if there isn't a better way to deal with it.

Reply:
Seems we should return `Array<valueof<T> | null>`.

In the case where T is an array of values, would that throw an error, or cause a compiler warning?

Reply:
Looks like that'd result in `AS204: Basic type 'X' cannot be nullable.`. One of those unfortunate dynamicness-of-JS cases again it seems.

Reply:
Apart from that, including `null` if a child is `null` appears like one of those questionable API design decisions on the JS-side. Would actually expect that it deals with `null` values just like if it was an empty array. So, from my perspective, it's probably fine to break with the JS-way here, skip those `null`s and document it. Thoughts?

Reply:
Yeah I went ahead and did this. I put a comment in the `d.ts` file too about how it will ignore null entries.